### PR TITLE
specify pod name and hostname in indexed job

### PR DESCRIPTION
**Pull request by AliceZhang2016**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
In Indexed Job:
- The Pod name takes the form `$(job-name)-$(index)-$(random-string)`, which can be used for quickly identifying Pods for a specific index when listing pods or looking at logs.
- The Pod hostname takes the form `$(job-name)-$(index)` which can be used to address the Pod from others, when the Job is used in combination with a headless Service.

#### Which issue(s) this PR fixes:

kubernetes/enhancements#2630

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2214-indexed-job
```

### CODE REVIEW

**Comment by AliceZhang2016** [comment]
cc @alculquicondor

File: pkg/controller/controller_utils.go
Comment (alculquicondor) [inline pkg/controller/controller_utils.go:554]: This is missing the check for the existence of the annotation.

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:846]: why not just put the code here? The annotation required more lines, but this is one-liner.

File: pkg/controller/job/indexed_job_utils.go
Comment (alculquicondor) [inline pkg/controller/job/indexed_job_utils.go:210]: No need for `Itoa` if you are already using `Sprintf`. You can use `%d`

File: pkg/controller/job/job_controller_test.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:808]: use `cmp.Equal` for single values.
Also, maybe do this check inside `checkCompletionIndexesInPods`, which can be renamed to `checkIndexedJobPods`

File: pkg/controller/controller_utils.go
Comment (alculquicondor) [inline pkg/controller/controller_utils.go:31]: I'm not sure if adding this dependency is appropriate. But I'll leave it to OWNERS.

File: pkg/controller/job/job_controller.go
Comment (AliceZhang2016) [reply pkg/controller/job/job_controller.go:846]: thought of the uniform of the code. I'll modify to one-line if it is better

File: pkg/controller/controller_utils.go
Comment (AliceZhang2016) [reply pkg/controller/controller_utils.go:554]: The annotation is one part of struct`PodTemplateSpec.ObjectMeta` and it'll be defaulted as `nil` if it is not given value, why it may not exist?

Comment (alculquicondor) [reply pkg/controller/controller_utils.go:554]: but if it doesn't exist, with this could you would get as prefix: `job-name--` (double dash)

File: pkg/controller/job/job_controller_test.go
Comment (AliceZhang2016) [reply pkg/controller/job/job_controller_test.go:808]: Done

File: pkg/controller/controller_utils.go
Comment (AliceZhang2016) [reply pkg/controller/controller_utils.go:554]: ah you mean the existence of the `batch.JobCompletionIndexAnnotationAlpha` in the `Annotations` map. You're right

File: pkg/controller/job/job_controller_test.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:798]: great, we have to add the same check in integration and E2E test :)

Comment (AliceZhang2016) [reply pkg/controller/job/job_controller_test.go:798]: done

File: pkg/controller/controller_utils.go
Comment (ahg-g) [inline pkg/controller/controller_utils.go:555]: we need to change the annotation name, drop the Alpha suffix completely, better to do that in https://github.com/kubernetes/kubernetes/pull/101292

File: pkg/controller/controller_utils.go
Comment (ahg-g) [reply pkg/controller/controller_utils.go:554]: Ideally we should check the mode too, is it easy to pass in?

File: pkg/controller/controller_utils.go
Comment (AliceZhang2016) [reply pkg/controller/controller_utils.go:555]: then maybe hold on this PR until #101292 merged, so that the new annotation name can be used here

Comment (AliceZhang2016) [reply pkg/controller/controller_utils.go:555]: changes of annotation name committed in #101292

**Comment by ahg-g** [comment]
can you please squash

**Comment by ahg-g** [comment]
> can you please squash

sorry, this was meant for the other PR :)

**Comment by AliceZhang2016** [comment]
@janetkuo  Could you please take a look at this PR? Thanks!

File: pkg/controller/job/job_controller.go
Comment (liggitt) [inline pkg/controller/job/job_controller.go:846]: is this valid for all combinations of valid job names and valid completionIndex?

Comment (ahg-g) [reply pkg/controller/job/job_controller.go:846]: We don't currently validate that the job name is a valid dns name, we should do that for jobs with `CompletionMode= IndexedCompletion`. @alculquicondor is there a maximum index? we should subtract the length from the max allowed DNs subdomain length like we do for cron job: https://github.com/kubernetes/kubernetes/blob/aa450787bff54c96b4588e7ddbfdf9bda89a882a/pkg/apis/batch/validation/validation.go#L208.

Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:846]: The completions field is an `int32`, so 10 would be the max number of digits.

@liggitt can we tighten the validation for an alpha feature?

Comment (liggitt) [reply pkg/controller/job/job_controller.go:846]: 10 + 1 for the dash, right?

you can do the validation when creating an indexed job, and when updating a non-indexed job to an indexed job

Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:846]: completionMode is immutable :)

Comment (liggitt) [reply pkg/controller/job/job_controller.go:846]: even better! validating on creation sgtm.

File: pkg/controller/controller_utils.go
Comment (janetkuo) [reply pkg/controller/controller_utils.go:31]: This is fine

File: pkg/controller/controller_utils.go
Comment (AliceZhang2016) [reply pkg/controller/controller_utils.go:554]: sorry for just notice this review as the conversation resolved and hidden.
I don't know how to get the job mode information from the `runtime.Object`, otherwise we need to add additional parameter to it.
Since the `batch.JobCompletionIndexAnnotationAlpha` is only added when it is indexed job (the mode is also checked there), do we really have to add the mode check here?

Comment (alculquicondor) [reply pkg/controller/controller_utils.go:554]: It would be so much effort to pass the Job object through that I don't think it's worth. Besides, this package is used for other controllers.

Comment (ahg-g) [reply pkg/controller/controller_utils.go:554]: sg

File: pkg/controller/controller_utils.go
Comment (liggitt) [inline pkg/controller/controller_utils.go:551]: getPodsPrefix handles the length being exceeded by trimming off the trailing dash. what happens if `prefix` is already at its max and we then append an index?

File: pkg/controller/controller_utils.go
Comment (liggitt) [inline pkg/controller/controller_utils.go:555]: it's strange to add job-specific logic to a generic controller helper method. the other place that tweaks pod names for an index (statefulset) does so with its own method:

https://github.com/kubernetes/kubernetes/blob/a5cf298a95b4a60789a6ba1caab931bdc980aa2d/pkg/controller/statefulset/stateful_set_utils.go#L249-L256

Comment (liggitt) [reply pkg/controller/controller_utils.go:555]: letting the CreatePodsWithControllerRef caller indicate a pod name suffix might make more sense than baking job-specific logic into this method

File: pkg/controller/job/job_controller.go
Comment (liggitt) [reply pkg/controller/job/job_controller.go:846]: validation added in https://github.com/kubernetes/kubernetes/pull/101292

File: pkg/controller/job/job_controller.go
Comment (liggitt) [inline pkg/controller/job/job_controller.go:843]: on line 837 above, won't that cause an out of bounds panic if `indexesToAdd = indexesToAdd[1:]` truncates `indexesToAdd` to `[]` and the loop runs another time? shouldn't that be `if len(indexesToAdd) > 0 {`?

File: pkg/controller/controller_utils.go
Comment (AliceZhang2016) [reply pkg/controller/controller_utils.go:551]: After experiments, it will be truncated later and append a random 5-character string to finally contain at most 63 characters, which means the appended index will also be discarded. I think better to truncate the prefix (63-5-len(maxIndex)-2(two dashes)) here to reserve the space for adding the suffix(`$index`)

File: pkg/controller/controller_utils.go
Comment (AliceZhang2016) [reply pkg/controller/controller_utils.go:555]: `CreatePodsWithControllerRef` call `CreatePods`, and the function `createPods` will record an event which contains the created pod name. I think it’s better to add index suffix before outputting the event.

Or if we don’t use the generic method `CreatePods`, we can write something like statefulset, but that may need much effort?

Comment (alculquicondor) [reply pkg/controller/controller_utils.go:555]: > letting the CreatePodsWithControllerRef caller indicate a pod name suffix might make more sense

I agree, but if a lot of callers have to change and the owners of the file are fine with this change as is, I think it's acceptable.

Just in case, @AliceZhang2016 could you check if adding a `CreatePodsWithControllerRefAndPrefix` to this file would be a lot of effort. I was kind of thinking in that direction when I created this cleanup #101156

File: pkg/controller/job/job_controller.go
Comment (liggitt) [reply pkg/controller/job/job_controller.go:843]: those are pretty separated, it's easy to imagine that logic changing. if the zero-length check is correct, it seems more robust

Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:843]: sounds good, will change in #98238

File: pkg/controller/controller_utils.go
Comment (alculquicondor) [inline pkg/controller/controller_utils.go:547]: just to avoid this change, hopefully we can merge #101156 first

@soltysh can you approve? :)

File: pkg/controller/controller_utils.go
Comment (alculquicondor) [inline pkg/controller/controller_utils.go:582]: I think we should be good without this change

File: pkg/controller/controller_utils.go
Comment (alculquicondor) [inline pkg/controller/controller_utils.go:571]: I think we can name the variable just `generateName`.

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:853]: shouldn't this be `len(generateNamePrefix) > names.MaxGeneratedNameLenght`?

And below, you probably need to cut the job name to make space for the index.

Better extract this in a utility function in `indexed_job_utils.go` and add a test for it.

Comment (AliceZhang2016) [reply pkg/controller/job/job_controller.go:853]: We already cut the job name as a part of generateName for make space for the index. Anywhere else needed to cut job name for index?

File: pkg/controller/controller_utils.go
Comment (AliceZhang2016) [reply pkg/controller/controller_utils.go:582]: In that case, `CreatePods` needs to pass `nil` into it, and seems no functionality difference between of`CreatePods` and `CreatePodsWithControllerRef`?

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:853]: I'm not sure the code is correct for all sizes. Extract the code in a function and add a test.

File: pkg/controller/controller_utils.go
Comment (AliceZhang2016) [reply pkg/controller/controller_utils.go:571]: Done.

File: pkg/controller/job/job_controller.go
Comment (AliceZhang2016) [reply pkg/controller/job/job_controller.go:853]: Done.

File: pkg/controller/job/indexed_job_utils.go
Comment (alculquicondor) [inline pkg/controller/job/indexed_job_utils.go:210]: nit: `podGenerateNameWithIndex`

we avoid the word `get` in golang.

File: pkg/controller/job/indexed_job_utils_test.go
Comment (alculquicondor) [inline pkg/controller/job/indexed_job_utils_test.go:262]: Add a test with a longer index, like `123456789`, and a `jobName` that is close to `MaxGeneratedNameLength` (0, 1, -1 characters of difference).

File: pkg/controller/job/indexed_job_utils.go
Comment (AliceZhang2016) [reply pkg/controller/job/indexed_job_utils.go:210]: Done.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: pkg/controller/controller_utils.go
Comment (AliceZhang2016) [reply pkg/controller/controller_utils.go:576]: Done.

File: pkg/controller/job/indexed_job_utils.go
Comment (AliceZhang2016) [reply pkg/controller/job/indexed_job_utils.go:222]: I can't figure out how to make it simpler. Any code suggestions about it?

File: pkg/controller/controller_utils.go
Comment (AliceZhang2016) [reply pkg/controller/controller_utils.go:530]: Let the `CreatePodsWithControllerRef` call this method to share codes. Also add a unit test about this method.

File: pkg/controller/statefulset/stateful_set_utils.go
Comment (AliceZhang2016) [reply pkg/controller/statefulset/stateful_set_utils.go]: roll back this change.

File: pkg/controller/controller_utils.go
Comment (alculquicondor) [inline pkg/controller/controller_utils.go:518]: wait a second.... nothing is using this function. I'll get rid of it in a follow up.

But since `CreatePods` is going away, it makes sense to rename `CreatePodsWithControllerRef` to simply `CreatePods`. Then, `CreatePodsWithControllerRefAndPrefix` can simply be `CreatePodsWithPrefix`. But I can do that in the follow up too.

**Review by alculquicondor — COMMENTED** [review]
/lgtm

File: pkg/controller/controller_utils_test.go
Comment (soltysh) [inline pkg/controller/controller_utils_test.go:32]: Nit `batchv1` is preferred.

File: pkg/controller/job/indexed_job_utils.go
Comment (soltysh) [reply pkg/controller/job/indexed_job_utils.go:222]:
```golang
rateNamePrefix := jobName +  "-" + strconv.Itoa(index) + "-"
```
:wink:

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

File: pkg/controller/job/indexed_job_utils.go
Comment (AliceZhang2016) [reply pkg/controller/job/indexed_job_utils.go:222]: If change to https://github.com/kubernetes/kubernetes/pull/101601#discussion_r634699571, then `appendIndex` in the following `if` block will be more complicated to represent.
I'd keep the current code.

File: pkg/controller/controller_utils_test.go
Comment (AliceZhang2016) [reply pkg/controller/controller_utils_test.go]: Done. Thanks for reviewing : )

**Comment by AliceZhang2016** [comment]
@liggitt Kindly ping to see if the PR can be approved  : ) Need approval from storage owners as the PR changed a constant there.

File: pkg/controller/controller_utils.go
Comment (liggitt) [inline pkg/controller/controller_utils.go:452]: the docs are confusing... what prefix is set?

Comment (AliceZhang2016) [reply pkg/controller/controller_utils.go]: Update method name to `CreatePodsWithControllerRefAndGenerateName `.
Thanks for reviewing : )

**Pull request merged**
